Task manager top level repair tasks

Deexie opened the pull request:
The PR introduces top level repair tasks representing repair and node operations
performed with repair. The actions performed as a part of these operations are 
moved to corresponding tasks' run methods.

Also a small change to repair module is added.

Deexie commented on repair/repair_task.hh:
```diff
@@ -0,0 +1,33 @@
+/*
```
#11851

Deexie commented on main.cc:
```diff
@@ -1320,7 +1320,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             // both)
```
#11851

Deexie commented on repair/repair.cc:
```diff
@@ -167,7 +167,7 @@ static std::vector<sstring> list_column_families(const replica::database& db, co
 }
```
#11851

Deexie commented on repair/repair_task.hh:
```diff
@@ -17,6 +17,13 @@ public:
         : tasks::task_manager::task::impl(module, id, sequence_number, std::move(keyspace), std::move(table), std::move(type), std::move(entity), parent_id) {
```
#11851

Deexie commented on repair/repair.cc:
```diff
@@ -370,32 +370,35 @@ void tracker::done(repair_uniq_id id, bool succeeded) {
     }
```
#11851

Deexie commented on repair/repair.cc:
```diff
@@ -386,7 +386,7 @@ repair_status repair_module::get(int id) const {
 }
```
#11851

Deexie commented on repair/repair.cc:
```diff
@@ -348,7 +348,6 @@ float node_ops_metrics::repair_finished_percentage() {
 repair_module::repair_module(tasks::task_manager& tm, repair_service& rs, size_t max_repair_memory) noexcept
```
#11851

Deexie commented on repair/repair.cc:
```diff
@@ -348,7 +348,6 @@ float node_ops_metrics::repair_finished_percentage() {
 repair_module::repair_module(tasks::task_manager& tm, repair_service& rs, size_t max_repair_memory) noexcept
     : tasks::task_manager::module(tm, "repair")
     , _rs(rs)
-    , _shutdown(false)
```
#11851

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2815/

bhalevy commented:
@Deexie does this PR need to be rebased?

bhalevy commented on tasks/task_manager.hh:
```diff
@@ -126,6 +128,16 @@ public:
+                        });
+                    });
+                }
                 return make_ready_future<>();
```
This could be simpler written as a coroutine.

bhalevy replied:
Note that instead of `with_semaphore` you can use get_units to hold the semaphore while iterating over `_children`.

bhalevy commented on tasks/task_manager.hh:
```diff
@@ -337,17 +352,22 @@ public:
                     if (auto it = all_tasks.find(id); it != all_tasks.end()) {
-                        it->second->add_child(std::move(task));
-                        return make_ready_future<uint64_t>(it->second->get_sequence_number());
+                        return it->second->add_child(std::move(task), abort).then([it] {
```
Since you have access to the parent task here, via `it->second` there's no need to push `abort` to `add_child`.
Assuming we have a `task::abort_requested()` method that peers into _impl->_as,
we can call it here after adding the child.

bhalevy commented on repair/repair.cc:
```diff
@@ -936,6 +936,11 @@ struct repair_options {
 
+static future<tasks::task_manager::task_ptr> start_repair_task(tasks::task_manager::task::task_impl_ptr task_impl_ptr, shared_ptr<repair_module> module, tasks::task_info pd = {}) {
+    auto task = co_await module->make_task(std::move(task_impl_ptr), pd);
+    task->start();
```
start should handle the aborted case where the task is already aborted at this point (if the parent task is already aborted), so it doesn't set the task state to `running`.
We should probably run the "monitor" fiber that calls `unregister_task` after the ttl expires,
but not call `run_to_completion`.

Deexie replied:
Do we really need to check parent task? I think that `this->abort_requested()` would be enough:
- if parent is aborted, then task is created -> task is aborted
- if task is created, then parent is aborted, then task is started -> task abortion is populated (and maybe done).

Also, then `start()` would be able to still return `void` and not `future<>`, which may be useful in some cases.

bhalevy replied:
I didn't see we need to check the parent task :)
You're right, if the parent task is already aborted, make_task will return a task that's already aborted as well.
I'm saying that task->start() should handle this case and e.g. refrain from calling `run_to_completion()`.

Deexie replied:
Done: #11995

bhalevy reviewed (changes requested):
General comment: the PR subject should be about repair (top level tasks).

Deexie commented:
- changed `abort()`  and `add_child()` to coroutines
- `_children_access_semaphore` -> `_children_semaphore`
- deleted abort parameter from `add_child` and added `task::abort_requested()`
- tasks that were aborted before started do not run
- tasks are aborted on shutdown
- top level repair tasks are abortable
- added noexcept to task_impl constructor

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2938/

Deexie commented:
- dropped commits concerning aborting tasks to reduce chaos; 
the commits together with other abort related changes would be 
submitted as a separate PR.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2980/

bhalevy commented on repair/repair.cc:
```diff
@@ -1106,12 +1111,23 @@ int repair_service::do_repair_start(sstring keyspace, std::unordered_map<sstring
     if (cfs.empty()) {
         rlogger.info("repair[{}]: completed successfully: no tables to repair", id.uuid());
-        return id.id;
+        return make_ready_future<int>(id.id);
```
nit: just co_return it

bhalevy commented on repair/repair.cc:
```diff
@@ -1106,12 +1111,23 @@ int repair_service::do_repair_start(sstring keyspace, std::unordered_map<sstring
-    (void)get_repair_module().run(id, [this, &db, id, keyspace = std::move(keyspace),
-            cfs = std::move(cfs), ranges = std::move(ranges), options = std::move(options), ignore_nodes = std::move(ignore_nodes)] () mutable {
+    auto task_impl_ptr = std::make_unique<user_requested_repair_task_impl>(_repair_module, id, std::move(keyspace), format("{}", streaming::stream_reason::repair), "", std::move(cfs), std::move(ranges), std::move(options.hosts), std::move(options.data_centers), std::move(ignore_nodes));
+    return start_repair_task(std::move(task_impl_ptr), _repair_module).then([id] (auto task) {
```
would be simpler as a coroutine

bhalevy reviewed (changes requested):
minor changes requested

Deexie commented:
- coroutinized `repair_service::do_repair_start`
- changed order of `repair_service::do_sync_data_using_repair` and `data_sync_repair_task_impl::run()`

bhalevy reviewed (approved):
LGTM.

Note: there might be conflicts with https://github.com/scylladb/scylladb/pull/11994
Since the latter is fixing an issue, it should be merged in higher priority but if its review process will drag, we can merge this one first and I'll rebase #11994 on top of it.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3009/

denesb commented:
@Deexie conflicts, please rebase.

Deexie commented:
- rebased

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3126/
